homeassistant/components/broadlink/remote.py

```diff
@@ -0,0 +1,418 @@
+"""Support for Broadlink IR/RF remotes."""
+import asyncio
+from base64 import b64encode
+from binascii import hexlify
+from datetime import timedelta
+from ipaddress import ip_address
+from itertools import product, takewhile, zip_longest
+import json
+import logging
+import socket
+
+import voluptuous as vol
+
+from homeassistant.components.remote import (
+    ATTR_ALTERNATIVE,
+    ATTR_COMMAND,
+    ATTR_DELAY_SECS,
+    ATTR_DEVICE,
+    ATTR_NUM_REPEATS,
+    ATTR_TIMEOUT,
+    DEFAULT_DELAY_SECS,
+    DEFAULT_NUM_REPEATS,
+    DOMAIN as COMPONENT,
+    PLATFORM_SCHEMA,
+    SUPPORT_LEARN_COMMAND,
+    RemoteDevice,
+)
+from homeassistant.const import CONF_HOST, CONF_MAC, CONF_NAME, CONF_TIMEOUT
+from homeassistant.exceptions import PlatformNotReady
+import homeassistant.helpers.config_validation as cv
+from homeassistant.util.dt import utcnow
+
+from . import DOMAIN, data_packet, hostname, mac_address
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_METADATA = ".metadata"
+
+DEFAULT_LEARNING_TIMEOUT = 20
+DEFAULT_NAME = "Broadlink"
+DEFAULT_PORT = 80
+DEFAULT_RETRY = 3
+DEFAULT_TIMEOUT = 5
+
+SCAN_INTERVAL = timedelta(minutes=2)
+
+COMMAND_SCHEMA = vol.Schema(
+    {
+        vol.All(str, vol.Length(min=1)): {  # Device
+            CONF_METADATA: dict,
+            vol.All(str, vol.Length(min=1)): vol.Any(  # Command
+                vol.All(str, vol.Length(min=1)),
+                vol.All([vol.All(str, vol.Length(min=1))], vol.Length(min=2, max=2)),
+            ),
+        }
+    }
+)
+
+TOGGLE_SCHEMA = vol.Schema({cv.string: vol.In([0, 1])})
+
+MINIMUM_SERVICE_SCHEMA = vol.Schema(
+    {
+        vol.Required(ATTR_COMMAND): vol.All(
+            cv.ensure_list,
+            [vol.All(cv.string, vol.Length(min=1), vol.NotIn([CONF_METADATA]))],
+            vol.Length(min=1),
+        ),
+        vol.Required(ATTR_DEVICE): vol.All(cv.string, vol.Length(min=1)),
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_HOST): vol.All(vol.Any(hostname, ip_address), cv.string),
+        vol.Required(CONF_MAC): mac_address,
+        vol.Optional(CONF_TIMEOUT, default=DEFAULT_TIMEOUT): cv.positive_int,
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+    }
+)
+
+
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
+    """Set up the Broadlink remote."""
+    import broadlink
+
+    host = config.get(CONF_HOST)
```

Comment:
Please use `dict[key]` for required config keys and keys with default config schema values.

Reply:
Done.